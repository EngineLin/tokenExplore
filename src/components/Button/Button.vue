<script>
export default {
  name: 'button',
  props: {
    buttonValue: {
      type: String,
      default: '按鈕',
    },
  },
  methods: {
    linkToOtherPage() {
      const linkValue = this.buttonValue;
      if (typeof linkValue !== 'string' || linkValue === '按鈕') return;

      this.$emit('handleLinkToOtherPage', linkValue);
    },
  },
};
</script>

<template lang="pug">
#app
  button.style-button(@click='linkToOtherPage') {{ buttonValue }}
</template>

<style lang="sass" scoped>
.style-button
  width: 100px
  height: 50px
  background-color: #fdc7ff
  border-radius: 3px
  box-shadow: 0px 0px
  &:hover
    cursor: pointer
</style>
