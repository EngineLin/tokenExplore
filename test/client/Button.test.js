import { mount } from 'vue-test-utils'
import Button from '../../src/components/Button/Button.vue'

/**
 * 這邊可以思考一下測試的時候基本需要測試的項目是什麼？
 * 慢慢累積經驗。
 * 1. props
 * 2. methods and $emit
 * 3. computed and watch
 * 4. template
 * 5. slot
 */
describe('Button Component 測試', () => {
  let cmp;
  beforeEach(() => {
    cmp = mount(Button);
  });

  describe('about Constructure', () => {
    it('snapshot:', () => {
      expect(cmp.element).toMatchSnapshot();
    });
  });

  describe('about Props:', () => {
    it("預設是 '按鈕'", () => {
      expect(cmp.props().buttonValue).toBe('按鈕');
    });
    it("當 Props 帶入 '註冊' 時會成功帶到 template 上", () => {
      cmp.setProps({ buttonValue: '註冊' });
      expect(cmp.find('button').text()).toBe('註冊');
    });
  });

  describe('about Methods:', () => {
    it('當 button 被點擊時，linkToOtherPage 被呼叫', () => {
      const mock = jest.fn();
      cmp.setMethods({ linkToOtherPage: mock });
      cmp.find('button').trigger('click');
      expect(mock).toBeCalled();
    });
    it("當 props.buttonValue 的類別不是 'string'，$emit.handleLinkToOtherPage 不被呼叫", () => {
      const mock = jest.fn();
      cmp.setProps({ buttonValue: 1 });
      cmp.vm.$on('handleLinkToOtherPage', mock);
      cmp.vm.linkToOtherPage();
      expect(mock).not.toBeCalled();
    });
    it("當 props.buttonValue 的內容是 '按鈕'，$emit.handleLinkToOtherPage 不被呼叫", () => {
      const mock = jest.fn();
      cmp.setProps({ buttonValue: '按鈕' });
      cmp.vm.$on('handleLinkToOtherPage', mock);
      cmp.vm.linkToOtherPage();
      expect(mock).not.toBeCalled();
    });
    it("當 props.buttonValue 的類別是 'string' 且不等於 '按鈕' 時，$emit.handleLinkToOtherPage 被呼叫", () => {
      const mock = jest.fn();
      cmp.setProps({ buttonValue: '註冊' });
      cmp.vm.$on('handleLinkToOtherPage', mock);
      cmp.vm.linkToOtherPage();
      expect(mock).toBeCalledWith('註冊');
    });
  });
});
