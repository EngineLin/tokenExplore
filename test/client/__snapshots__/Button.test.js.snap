// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component 測試 about Constructure snapshot: 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "element": <div
    id="app"
  >
    <button
      class="style-button"
    >
      按鈕
    </button>
  </div>,
  "isVueComponent": true,
  "options": Object {
    "attachedToDocument": false,
  },
  "update": [Function],
  "version": 2.5,
  "vm": VueComponent {
    "$_mountingOptionsSlots": undefined,
    "$_originalSlots": Object {},
    "$attrs": Object {},
    "$children": Array [],
    "$createElement": [Function],
    "$el": <div
      id="app"
    >
      <button
        class="style-button"
      >
        按鈕
      </button>
    </div>,
    "$emit": [Function],
    "$listeners": Object {},
    "$options": Object {
      "_Ctor": Object {
        "1": [Function],
      },
      "_base": [Function],
      "_parentElm": null,
      "_propKeys": Array [
        "buttonValue",
      ],
      "_refElm": null,
      "beforeCreate": Array [
        [Function],
      ],
      "components": Object {
        "transition": Object {
          "render": [Function],
        },
        "transition-group": Object {
          "render": [Function],
        },
      },
      "directives": Object {},
      "filters": Object {},
      "methods": Object {
        "linkToOtherPage": [Function],
      },
      "name": "button",
      "props": Object {
        "buttonValue": Object {
          "default": "按鈕",
          "type": [Function],
        },
      },
      "render": [Function],
      "staticRenderFns": Array [],
    },
    "$parent": undefined,
    "$refs": Object {},
    "$root": [Circular],
    "$scopedSlots": Object {},
    "$slots": Object {},
    "$vnode": undefined,
    "__emitted": Object {},
    "__emittedByOrder": Array [],
    "_c": [Function],
    "_data": Object {},
    "_directInactive": false,
    "_events": Object {},
    "_hasHookEvent": false,
    "_inactive": null,
    "_isBeingDestroyed": false,
    "_isDestroyed": false,
    "_isMounted": true,
    "_isVue": true,
    "_props": Object {
      "buttonValue": "按鈕",
    },
    "_renderProxy": VueComponent {
      "$_mountingOptionsSlots": undefined,
      "$_originalSlots": Object {},
      "$attrs": Object {},
      "$children": Array [],
      "$createElement": [Function],
      "$el": <div
        id="app"
      >
        <button
          class="style-button"
        >
          按鈕
        </button>
      </div>,
      "$emit": [Function],
      "$listeners": Object {},
      "$options": Object {
        "_Ctor": Object {
          "1": [Function],
        },
        "_base": [Function],
        "_parentElm": null,
        "_propKeys": Array [
          "buttonValue",
        ],
        "_refElm": null,
        "beforeCreate": Array [
          [Function],
        ],
        "components": Object {
          "transition": Object {
            "render": [Function],
          },
          "transition-group": Object {
            "render": [Function],
          },
        },
        "directives": Object {},
        "filters": Object {},
        "methods": Object {
          "linkToOtherPage": [Function],
        },
        "name": "button",
        "props": Object {
          "buttonValue": Object {
            "default": "按鈕",
            "type": [Function],
          },
        },
        "render": [Function],
        "staticRenderFns": Array [],
      },
      "$parent": undefined,
      "$refs": Object {},
      "$root": [Circular],
      "$scopedSlots": Object {},
      "$slots": Object {},
      "$vnode": undefined,
      "__emitted": Object {},
      "__emittedByOrder": Array [],
      "_c": [Function],
      "_data": Object {},
      "_directInactive": false,
      "_events": Object {},
      "_hasHookEvent": false,
      "_inactive": null,
      "_isBeingDestroyed": false,
      "_isDestroyed": false,
      "_isMounted": true,
      "_isVue": true,
      "_props": Object {
        "buttonValue": "按鈕",
      },
      "_renderProxy": [Circular],
      "_self": [Circular],
      "_staticTrees": null,
      "_uid": 0,
      "_vnode": VNode {
        "asyncFactory": undefined,
        "asyncMeta": undefined,
        "children": Array [
          VNode {
            "asyncFactory": undefined,
            "asyncMeta": undefined,
            "children": Array [
              VNode {
                "asyncFactory": undefined,
                "asyncMeta": undefined,
                "children": undefined,
                "componentInstance": undefined,
                "componentOptions": undefined,
                "context": undefined,
                "data": undefined,
                "elm": 按鈕,
                "fnContext": undefined,
                "fnOptions": undefined,
                "fnScopeId": undefined,
                "isAsyncPlaceholder": false,
                "isCloned": false,
                "isComment": false,
                "isOnce": false,
                "isRootInsert": false,
                "isStatic": false,
                "key": undefined,
                "ns": undefined,
                "parent": undefined,
                "raw": false,
                "tag": undefined,
                "text": "按鈕",
              },
            ],
            "componentInstance": undefined,
            "componentOptions": undefined,
            "context": [Circular],
            "data": Object {
              "on": Object {
                "click": [Function],
              },
              "staticClass": "style-button",
            },
            "elm": <button
              class="style-button"
            >
              按鈕
            </button>,
            "fnContext": undefined,
            "fnOptions": undefined,
            "fnScopeId": undefined,
            "isAsyncPlaceholder": false,
            "isCloned": false,
            "isComment": false,
            "isOnce": false,
            "isRootInsert": false,
            "isStatic": false,
            "key": undefined,
            "ns": undefined,
            "parent": undefined,
            "raw": false,
            "tag": "button",
            "text": undefined,
          },
        ],
        "componentInstance": undefined,
        "componentOptions": undefined,
        "context": [Circular],
        "data": Object {
          "attrs": Object {
            "id": "app",
          },
        },
        "elm": <div
          id="app"
        >
          <button
            class="style-button"
          >
            按鈕
          </button>
        </div>,
        "fnContext": undefined,
        "fnOptions": undefined,
        "fnScopeId": undefined,
        "isAsyncPlaceholder": false,
        "isCloned": false,
        "isComment": false,
        "isOnce": false,
        "isRootInsert": true,
        "isStatic": false,
        "key": undefined,
        "ns": undefined,
        "parent": undefined,
        "raw": false,
        "tag": "div",
        "text": undefined,
      },
      "_watcher": Watcher {
        "active": true,
        "cb": [Function],
        "deep": false,
        "depIds": Set {
          2,
        },
        "deps": Array [
          Dep {
            "id": 2,
            "subs": Array [
              [Circular],
            ],
          },
        ],
        "dirty": false,
        "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
        "getter": [Function],
        "id": 1,
        "lazy": false,
        "newDepIds": Set {},
        "newDeps": Array [],
        "sync": false,
        "user": false,
        "value": undefined,
        "vm": [Circular],
      },
      "_watchers": Array [
        Watcher {
          "active": true,
          "cb": [Function],
          "deep": false,
          "depIds": Set {
            2,
          },
          "deps": Array [
            Dep {
              "id": 2,
              "subs": Array [
                [Circular],
              ],
            },
          ],
          "dirty": false,
          "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
          "getter": [Function],
          "id": 1,
          "lazy": false,
          "newDepIds": Set {},
          "newDeps": Array [],
          "sync": false,
          "user": false,
          "value": undefined,
          "vm": [Circular],
        },
      ],
      "linkToOtherPage": [Function],
    },
    "_self": [Circular],
    "_staticTrees": null,
    "_uid": 0,
    "_vnode": VNode {
      "asyncFactory": undefined,
      "asyncMeta": undefined,
      "children": Array [
        VNode {
          "asyncFactory": undefined,
          "asyncMeta": undefined,
          "children": Array [
            VNode {
              "asyncFactory": undefined,
              "asyncMeta": undefined,
              "children": undefined,
              "componentInstance": undefined,
              "componentOptions": undefined,
              "context": undefined,
              "data": undefined,
              "elm": 按鈕,
              "fnContext": undefined,
              "fnOptions": undefined,
              "fnScopeId": undefined,
              "isAsyncPlaceholder": false,
              "isCloned": false,
              "isComment": false,
              "isOnce": false,
              "isRootInsert": false,
              "isStatic": false,
              "key": undefined,
              "ns": undefined,
              "parent": undefined,
              "raw": false,
              "tag": undefined,
              "text": "按鈕",
            },
          ],
          "componentInstance": undefined,
          "componentOptions": undefined,
          "context": [Circular],
          "data": Object {
            "on": Object {
              "click": [Function],
            },
            "staticClass": "style-button",
          },
          "elm": <button
            class="style-button"
          >
            按鈕
          </button>,
          "fnContext": undefined,
          "fnOptions": undefined,
          "fnScopeId": undefined,
          "isAsyncPlaceholder": false,
          "isCloned": false,
          "isComment": false,
          "isOnce": false,
          "isRootInsert": false,
          "isStatic": false,
          "key": undefined,
          "ns": undefined,
          "parent": undefined,
          "raw": false,
          "tag": "button",
          "text": undefined,
        },
      ],
      "componentInstance": undefined,
      "componentOptions": undefined,
      "context": [Circular],
      "data": Object {
        "attrs": Object {
          "id": "app",
        },
      },
      "elm": <div
        id="app"
      >
        <button
          class="style-button"
        >
          按鈕
        </button>
      </div>,
      "fnContext": undefined,
      "fnOptions": undefined,
      "fnScopeId": undefined,
      "isAsyncPlaceholder": false,
      "isCloned": false,
      "isComment": false,
      "isOnce": false,
      "isRootInsert": true,
      "isStatic": false,
      "key": undefined,
      "ns": undefined,
      "parent": undefined,
      "raw": false,
      "tag": "div",
      "text": undefined,
    },
    "_watcher": Watcher {
      "active": true,
      "cb": [Function],
      "deep": false,
      "depIds": Set {
        2,
      },
      "deps": Array [
        Dep {
          "id": 2,
          "subs": Array [
            [Circular],
          ],
        },
      ],
      "dirty": false,
      "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
      "getter": [Function],
      "id": 1,
      "lazy": false,
      "newDepIds": Set {},
      "newDeps": Array [],
      "sync": false,
      "user": false,
      "value": undefined,
      "vm": [Circular],
    },
    "_watchers": Array [
      Watcher {
        "active": true,
        "cb": [Function],
        "deep": false,
        "depIds": Set {
          2,
        },
        "deps": Array [
          Dep {
            "id": 2,
            "subs": Array [
              [Circular],
            ],
          },
        ],
        "dirty": false,
        "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
        "getter": [Function],
        "id": 1,
        "lazy": false,
        "newDepIds": Set {},
        "newDeps": Array [],
        "sync": false,
        "user": false,
        "value": undefined,
        "vm": [Circular],
      },
    ],
    "linkToOtherPage": [Function],
  },
  "vnode": VNode {
    "asyncFactory": undefined,
    "asyncMeta": undefined,
    "children": Array [
      VNode {
        "asyncFactory": undefined,
        "asyncMeta": undefined,
        "children": Array [
          VNode {
            "asyncFactory": undefined,
            "asyncMeta": undefined,
            "children": undefined,
            "componentInstance": undefined,
            "componentOptions": undefined,
            "context": undefined,
            "data": undefined,
            "elm": 按鈕,
            "fnContext": undefined,
            "fnOptions": undefined,
            "fnScopeId": undefined,
            "isAsyncPlaceholder": false,
            "isCloned": false,
            "isComment": false,
            "isOnce": false,
            "isRootInsert": false,
            "isStatic": false,
            "key": undefined,
            "ns": undefined,
            "parent": undefined,
            "raw": false,
            "tag": undefined,
            "text": "按鈕",
          },
        ],
        "componentInstance": undefined,
        "componentOptions": undefined,
        "context": VueComponent {
          "$_mountingOptionsSlots": undefined,
          "$_originalSlots": Object {},
          "$attrs": Object {},
          "$children": Array [],
          "$createElement": [Function],
          "$el": <div
            id="app"
          >
            <button
              class="style-button"
            >
              按鈕
            </button>
          </div>,
          "$emit": [Function],
          "$listeners": Object {},
          "$options": Object {
            "_Ctor": Object {
              "1": [Function],
            },
            "_base": [Function],
            "_parentElm": null,
            "_propKeys": Array [
              "buttonValue",
            ],
            "_refElm": null,
            "beforeCreate": Array [
              [Function],
            ],
            "components": Object {
              "transition": Object {
                "render": [Function],
              },
              "transition-group": Object {
                "render": [Function],
              },
            },
            "directives": Object {},
            "filters": Object {},
            "methods": Object {
              "linkToOtherPage": [Function],
            },
            "name": "button",
            "props": Object {
              "buttonValue": Object {
                "default": "按鈕",
                "type": [Function],
              },
            },
            "render": [Function],
            "staticRenderFns": Array [],
          },
          "$parent": undefined,
          "$refs": Object {},
          "$root": [Circular],
          "$scopedSlots": Object {},
          "$slots": Object {},
          "$vnode": undefined,
          "__emitted": Object {},
          "__emittedByOrder": Array [],
          "_c": [Function],
          "_data": Object {},
          "_directInactive": false,
          "_events": Object {},
          "_hasHookEvent": false,
          "_inactive": null,
          "_isBeingDestroyed": false,
          "_isDestroyed": false,
          "_isMounted": true,
          "_isVue": true,
          "_props": Object {
            "buttonValue": "按鈕",
          },
          "_renderProxy": VueComponent {
            "$_mountingOptionsSlots": undefined,
            "$_originalSlots": Object {},
            "$attrs": Object {},
            "$children": Array [],
            "$createElement": [Function],
            "$el": <div
              id="app"
            >
              <button
                class="style-button"
              >
                按鈕
              </button>
            </div>,
            "$emit": [Function],
            "$listeners": Object {},
            "$options": Object {
              "_Ctor": Object {
                "1": [Function],
              },
              "_base": [Function],
              "_parentElm": null,
              "_propKeys": Array [
                "buttonValue",
              ],
              "_refElm": null,
              "beforeCreate": Array [
                [Function],
              ],
              "components": Object {
                "transition": Object {
                  "render": [Function],
                },
                "transition-group": Object {
                  "render": [Function],
                },
              },
              "directives": Object {},
              "filters": Object {},
              "methods": Object {
                "linkToOtherPage": [Function],
              },
              "name": "button",
              "props": Object {
                "buttonValue": Object {
                  "default": "按鈕",
                  "type": [Function],
                },
              },
              "render": [Function],
              "staticRenderFns": Array [],
            },
            "$parent": undefined,
            "$refs": Object {},
            "$root": [Circular],
            "$scopedSlots": Object {},
            "$slots": Object {},
            "$vnode": undefined,
            "__emitted": Object {},
            "__emittedByOrder": Array [],
            "_c": [Function],
            "_data": Object {},
            "_directInactive": false,
            "_events": Object {},
            "_hasHookEvent": false,
            "_inactive": null,
            "_isBeingDestroyed": false,
            "_isDestroyed": false,
            "_isMounted": true,
            "_isVue": true,
            "_props": Object {
              "buttonValue": "按鈕",
            },
            "_renderProxy": [Circular],
            "_self": [Circular],
            "_staticTrees": null,
            "_uid": 0,
            "_vnode": [Circular],
            "_watcher": Watcher {
              "active": true,
              "cb": [Function],
              "deep": false,
              "depIds": Set {
                2,
              },
              "deps": Array [
                Dep {
                  "id": 2,
                  "subs": Array [
                    [Circular],
                  ],
                },
              ],
              "dirty": false,
              "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
              "getter": [Function],
              "id": 1,
              "lazy": false,
              "newDepIds": Set {},
              "newDeps": Array [],
              "sync": false,
              "user": false,
              "value": undefined,
              "vm": [Circular],
            },
            "_watchers": Array [
              Watcher {
                "active": true,
                "cb": [Function],
                "deep": false,
                "depIds": Set {
                  2,
                },
                "deps": Array [
                  Dep {
                    "id": 2,
                    "subs": Array [
                      [Circular],
                    ],
                  },
                ],
                "dirty": false,
                "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
                "getter": [Function],
                "id": 1,
                "lazy": false,
                "newDepIds": Set {},
                "newDeps": Array [],
                "sync": false,
                "user": false,
                "value": undefined,
                "vm": [Circular],
              },
            ],
            "linkToOtherPage": [Function],
          },
          "_self": [Circular],
          "_staticTrees": null,
          "_uid": 0,
          "_vnode": [Circular],
          "_watcher": Watcher {
            "active": true,
            "cb": [Function],
            "deep": false,
            "depIds": Set {
              2,
            },
            "deps": Array [
              Dep {
                "id": 2,
                "subs": Array [
                  [Circular],
                ],
              },
            ],
            "dirty": false,
            "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
            "getter": [Function],
            "id": 1,
            "lazy": false,
            "newDepIds": Set {},
            "newDeps": Array [],
            "sync": false,
            "user": false,
            "value": undefined,
            "vm": [Circular],
          },
          "_watchers": Array [
            Watcher {
              "active": true,
              "cb": [Function],
              "deep": false,
              "depIds": Set {
                2,
              },
              "deps": Array [
                Dep {
                  "id": 2,
                  "subs": Array [
                    [Circular],
                  ],
                },
              ],
              "dirty": false,
              "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
              "getter": [Function],
              "id": 1,
              "lazy": false,
              "newDepIds": Set {},
              "newDeps": Array [],
              "sync": false,
              "user": false,
              "value": undefined,
              "vm": [Circular],
            },
          ],
          "linkToOtherPage": [Function],
        },
        "data": Object {
          "on": Object {
            "click": [Function],
          },
          "staticClass": "style-button",
        },
        "elm": <button
          class="style-button"
        >
          按鈕
        </button>,
        "fnContext": undefined,
        "fnOptions": undefined,
        "fnScopeId": undefined,
        "isAsyncPlaceholder": false,
        "isCloned": false,
        "isComment": false,
        "isOnce": false,
        "isRootInsert": false,
        "isStatic": false,
        "key": undefined,
        "ns": undefined,
        "parent": undefined,
        "raw": false,
        "tag": "button",
        "text": undefined,
      },
    ],
    "componentInstance": undefined,
    "componentOptions": undefined,
    "context": VueComponent {
      "$_mountingOptionsSlots": undefined,
      "$_originalSlots": Object {},
      "$attrs": Object {},
      "$children": Array [],
      "$createElement": [Function],
      "$el": <div
        id="app"
      >
        <button
          class="style-button"
        >
          按鈕
        </button>
      </div>,
      "$emit": [Function],
      "$listeners": Object {},
      "$options": Object {
        "_Ctor": Object {
          "1": [Function],
        },
        "_base": [Function],
        "_parentElm": null,
        "_propKeys": Array [
          "buttonValue",
        ],
        "_refElm": null,
        "beforeCreate": Array [
          [Function],
        ],
        "components": Object {
          "transition": Object {
            "render": [Function],
          },
          "transition-group": Object {
            "render": [Function],
          },
        },
        "directives": Object {},
        "filters": Object {},
        "methods": Object {
          "linkToOtherPage": [Function],
        },
        "name": "button",
        "props": Object {
          "buttonValue": Object {
            "default": "按鈕",
            "type": [Function],
          },
        },
        "render": [Function],
        "staticRenderFns": Array [],
      },
      "$parent": undefined,
      "$refs": Object {},
      "$root": [Circular],
      "$scopedSlots": Object {},
      "$slots": Object {},
      "$vnode": undefined,
      "__emitted": Object {},
      "__emittedByOrder": Array [],
      "_c": [Function],
      "_data": Object {},
      "_directInactive": false,
      "_events": Object {},
      "_hasHookEvent": false,
      "_inactive": null,
      "_isBeingDestroyed": false,
      "_isDestroyed": false,
      "_isMounted": true,
      "_isVue": true,
      "_props": Object {
        "buttonValue": "按鈕",
      },
      "_renderProxy": VueComponent {
        "$_mountingOptionsSlots": undefined,
        "$_originalSlots": Object {},
        "$attrs": Object {},
        "$children": Array [],
        "$createElement": [Function],
        "$el": <div
          id="app"
        >
          <button
            class="style-button"
          >
            按鈕
          </button>
        </div>,
        "$emit": [Function],
        "$listeners": Object {},
        "$options": Object {
          "_Ctor": Object {
            "1": [Function],
          },
          "_base": [Function],
          "_parentElm": null,
          "_propKeys": Array [
            "buttonValue",
          ],
          "_refElm": null,
          "beforeCreate": Array [
            [Function],
          ],
          "components": Object {
            "transition": Object {
              "render": [Function],
            },
            "transition-group": Object {
              "render": [Function],
            },
          },
          "directives": Object {},
          "filters": Object {},
          "methods": Object {
            "linkToOtherPage": [Function],
          },
          "name": "button",
          "props": Object {
            "buttonValue": Object {
              "default": "按鈕",
              "type": [Function],
            },
          },
          "render": [Function],
          "staticRenderFns": Array [],
        },
        "$parent": undefined,
        "$refs": Object {},
        "$root": [Circular],
        "$scopedSlots": Object {},
        "$slots": Object {},
        "$vnode": undefined,
        "__emitted": Object {},
        "__emittedByOrder": Array [],
        "_c": [Function],
        "_data": Object {},
        "_directInactive": false,
        "_events": Object {},
        "_hasHookEvent": false,
        "_inactive": null,
        "_isBeingDestroyed": false,
        "_isDestroyed": false,
        "_isMounted": true,
        "_isVue": true,
        "_props": Object {
          "buttonValue": "按鈕",
        },
        "_renderProxy": [Circular],
        "_self": [Circular],
        "_staticTrees": null,
        "_uid": 0,
        "_vnode": [Circular],
        "_watcher": Watcher {
          "active": true,
          "cb": [Function],
          "deep": false,
          "depIds": Set {
            2,
          },
          "deps": Array [
            Dep {
              "id": 2,
              "subs": Array [
                [Circular],
              ],
            },
          ],
          "dirty": false,
          "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
          "getter": [Function],
          "id": 1,
          "lazy": false,
          "newDepIds": Set {},
          "newDeps": Array [],
          "sync": false,
          "user": false,
          "value": undefined,
          "vm": [Circular],
        },
        "_watchers": Array [
          Watcher {
            "active": true,
            "cb": [Function],
            "deep": false,
            "depIds": Set {
              2,
            },
            "deps": Array [
              Dep {
                "id": 2,
                "subs": Array [
                  [Circular],
                ],
              },
            ],
            "dirty": false,
            "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
            "getter": [Function],
            "id": 1,
            "lazy": false,
            "newDepIds": Set {},
            "newDeps": Array [],
            "sync": false,
            "user": false,
            "value": undefined,
            "vm": [Circular],
          },
        ],
        "linkToOtherPage": [Function],
      },
      "_self": [Circular],
      "_staticTrees": null,
      "_uid": 0,
      "_vnode": [Circular],
      "_watcher": Watcher {
        "active": true,
        "cb": [Function],
        "deep": false,
        "depIds": Set {
          2,
        },
        "deps": Array [
          Dep {
            "id": 2,
            "subs": Array [
              [Circular],
            ],
          },
        ],
        "dirty": false,
        "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
        "getter": [Function],
        "id": 1,
        "lazy": false,
        "newDepIds": Set {},
        "newDeps": Array [],
        "sync": false,
        "user": false,
        "value": undefined,
        "vm": [Circular],
      },
      "_watchers": Array [
        Watcher {
          "active": true,
          "cb": [Function],
          "deep": false,
          "depIds": Set {
            2,
          },
          "deps": Array [
            Dep {
              "id": 2,
              "subs": Array [
                [Circular],
              ],
            },
          ],
          "dirty": false,
          "expression": "function () {
      vm._update(vm._render(), hydrating);
    }",
          "getter": [Function],
          "id": 1,
          "lazy": false,
          "newDepIds": Set {},
          "newDeps": Array [],
          "sync": false,
          "user": false,
          "value": undefined,
          "vm": [Circular],
        },
      ],
      "linkToOtherPage": [Function],
    },
    "data": Object {
      "attrs": Object {
        "id": "app",
      },
    },
    "elm": <div
      id="app"
    >
      <button
        class="style-button"
      >
        按鈕
      </button>
    </div>,
    "fnContext": undefined,
    "fnOptions": undefined,
    "fnScopeId": undefined,
    "isAsyncPlaceholder": false,
    "isCloned": false,
    "isComment": false,
    "isOnce": false,
    "isRootInsert": true,
    "isStatic": false,
    "key": undefined,
    "ns": undefined,
    "parent": undefined,
    "raw": false,
    "tag": "div",
    "text": undefined,
  },
}
`;
